Cosmic ray detection.

cmccully opened the pull request:
Full working version.

cmccully commented:
@eteq @astrofrog @crawfordsm @cdeil @larrybradley As per discussion offline, we decided that it would be best to make my cosmic ray detection code an affiliated package and not pull it into the core at this time. As such, I have ported it here and have renamed it. Would you be willing to do a quick code review on this? I think most of the hard work is done from discussion here: https://github.com/astropy/imageutils/pull/35. However, I thought it could still be useful to have it looked at one more time. Once we are happy, I will pull this in and would like to release this on PyPi. Thanks for all of your help with this.

P.S. Can someone turn on the Travis build for this repo?

cdeil commented:
@cmccully - Thanks!

Currently the repo is called `astro-scrappy` and the Python package is called `scrappy` here.

The name `scrappy` is already taken on the Python package index (see https://pypi.python.org/pypi/Scrappy/), so you can't release this package under that name.

My suggestion would also be to rename the repo ... I always find it confusing if people use different names (e.g. .... ah, better not point fingers here ...).

cdeil commented:
You have a lot of information duplicated in `README.md` and the package docstring in `scrappy/__init__.py`. In my experience that will always cause extra work and inconsistencies in the long run, so I'd suggest having the description only in one place and then referring to it from other places if needed.

cdeil commented on scrappy/scrappy.pyx:
```diff
@@ -0,0 +1,740 @@
+    return (crmask.astype(np.bool), cleanarr)
+
+
+def updatemask(np.ndarray[np.float32_t, ndim=2, mode='c', cast=True] data,
```
`updatemask` -> `update_mask`

(I think that's a PEP8 rule or at least common in Python, plus consistent with `clean_meanmask` below)

cmccully replied:
Done.

cdeil commented on scrappy/scrappy.pyx:
```diff
@@ -0,0 +1,740 @@
+
+cdef void clean_meanmask(float[:, ::1] cleanarr, bool[:, ::1] crmask,
+                         bool[:, ::1] mask, int nx, int ny,
+                         float backgroundlevel):
```
`background_level`?

cmccully replied:
Done.

cdeil commented on scrappy/scrappy.pyx:
```diff
@@ -0,0 +1,740 @@
+                        cleanarr[j, i] = val / wsum
+
+
+def gausskernel(float psffwhm, int kernsize):
```
This duplicates the [Gaussian2DKernel](http://astropy.readthedocs.org/en/latest/api/astropy.convolution.Gaussian2DKernel.html) in Astropy core.

Is is possible to avoid that duplication?
I guess the version in Astropy core would have to be moved to Cython, but then you could import and use it in this Cython file?

cdeil replied:
Really all these kernel functions, also Moffat, should go in Astropy, they are needed in other affiliated packages as well.

cmccully replied:
That is a fair point. They don't really need to be ported over to Cython as these calculations are small and only need to be done once. One concern I have is time scale. If I do a pull request to add Moffat to the astropy.convolution when would that get released? Also, some of the utils functions could be quite useful in the core, like the median filters etc (though they would need some modification). At one point we discussed leaving the duplication for now, but beginning to port over the functionality over time. Discussion?

cdeil replied:
@eteq or @astrofrog - Could you please update https://github.com/astropy/astropy/wiki/Release-Calendar ? When is Astropy 1.1 planned?

@cmccully - Do you care that people can use this package with the stable Astropy 1.1 now? Or would it be OK that it only works with the dev version for a few months until Astropy 1.1 is released?
What we've done in other packages (e.g. photutils) to make it work temporarily is to bundle some things from Astropy core in `extern` for a few months or a year, and the remove it and import from Astropy directly when we updated the requirement on a more recent stable Astropy version.

Moving things to Astropy core to avoid code duplication in affiliated packages is always quite a bit of extra work and discussion, but in the long run it's worth it.

larrybradley replied:
@cdeil  `Moffat1D` and `Moffat2D` models are already available in `astropy` (they've been there for a while, but previously under the names `Beta1D` and `Beta2D`).  Do you mean a `Cython` version of Moffat?

cdeil replied:
@larrybradley – I didn't know the Moffat models were there. @cmccully said that there's no speed concern here, so maybe can be used as-is, no need to move them to Cython?

cmccully replied:
The Moffat models are there, but the wrapper to make the kernel is not. It is not complicated, but just curious about the release schedule.

cmccully commented:
On the point about naming, this is tricky. I was aware that scrappy exists on PyPi. That is why the repo has a different name. One convention is that we could have the pip name be astro-scrappy and the package name be astro_scrappy. This makes the package name a little long, but would maybe less confusing. I am open to discussion on this, but would not be too happy about going back to the drawing board on names. There has already been a lot of email traffic about what to call this. If someone has a good suggestion for a name I am will consider it. 
What do other people think about this point?

cdeil commented:
My 2 cents: I hate it when it's `scrappy` here and `astro-scrappy` there and `astro_scrappy` there ... hard to Google and remember and confusing. Why not `astroscrappy` everywhere?
(contradicting myself here a bit, I guess technically PEP8 says it should be `astro_scrappy`).

That's kind of what we did here: https://github.com/astroplanners/astroplan

cmccully commented:
I would be ok with astroscrappy or astro_scrappy. I think pip automatically changes underscores to dashes: http://stackoverflow.com/questions/16385099/why-does-pip-convert-underscores-to-dashes. Other thoughts?

cmccully commented:
@eteq @astrofrog @crawfordsm @cdeil @larrybradley Any other comments before I pull this in and release it? I am leaning towards converting the name from astro-scrappy to astroscrappy. Any other comments on the name? 

As for the duplication with astropy-core, I think I would like to release this basically as is, but then do a 1.1 release with astropy 1.1 that removes the duplication. Does this sound reasonable?

Any other comments before I release this?

cdeil commented:
With this branch, I see one test error on my Macbook (with latest Mac OS X and clang):
https://gist.github.com/cdeil/054ad197f04e41456b2a#file-gistfile1-txt-L690

For the docs there's this warning:
```
/Users/deil/code/astro-scrappy/build/lib.macosx-10.10-x86_64-3.4/scrappy/__init__.py:docstring of scrappy:2: WARNING: Title underline too short.

Astro-SCRAPPY: The Speedy Cosmic Ray Annihilation Package in Python
===================================
/Users/deil/code/astro-scrappy/build/lib.macosx-10.10-x86_64-3.4/scrappy/__init__.py:docstring of scrappy:40: WARNING: Bullet list ends without a blank line; unexpected unindent.
```
and the `U` syntax markup in the docstring doesn't work:

![screen shot 2015-06-22 at 19 40 13](https://cloud.githubusercontent.com/assets/852409/8288447/8888db5c-1916-11e5-941b-8970d90b5374.png)

I think it would be nice if the repo is also renamed from `astro-scrappy` to `astroscrappy`, so that it's consistent everywhere. This will have to be done by an Astropy org owner, I think.

cmccully commented:
@cdeil Sorry about that. I had fixed that issue on my local machine, but forgot to push the update. I just pushed the update and removed the incorrect syntax markup. Can you test it again? Thanks.

cdeil commented:
@cmccully – Now tests pass.

One more think I'd suggest is to make the code Py 2/3 compatible without `2to3` (if it isn't already) and then set `use_2to3=False` here: https://github.com/astropy/astro-scrappy/blob/master/setup.py#L115

astrofrog commented:
Note that not everything needs to be done in this PR by the way - for example the 2to3 could probably be left to another PR. It would be good to set up appveyor too, and that's also something that can be done in another PR.

cmccully commented:
Fortunately I don't need 2to3 so I just changed that flag. Anything else, before I should release it?

mwcraig commented:
@astrofrog or @embray --could you flip the switch for this on travis?

astrofrog commented:
@mwcraig - done.

@cmccully - can you push a trivial commit (fixing text formatting or adding an empty space somewhere) to trigger Travis?

Once this PR is ready, I would suggest merging it then waiting a day or so before releasing, in case anyone wants to try it out in more detail.

astrofrog commented on scrappy/utils/__init__.py:
```diff
@@ -1,4 +1,17 @@
+           'optmed25', 'medfilt3', 'medfilt5', 'medfilt7',
+           'sepmedfilt3', 'sepmedfilt5', 'sepmedfilt7', 'sepmedfilt9',
+           'subsample', 'rebin', 'convolve', 'laplaceconvolve',
+           'dilate3', 'dilate5']
```
You can add a newline here (there is one missing) to trigger a new Travis build

cmccully commented:
@astrofrog I triggered the Travis build. 

Question: For some of my tests, I cross check my utility functions against scipy routines. I don't use scipy anywhere else. Should scipy still be considered a dependency? I added it to the travis file, but do I need to add it under the pip release requirements (or a requirements.txt file)? 

@astrofrog or @eteq do you have an opinion about astroscrappy vs astro-scrappy?

astrofrog commented:
@cmccully - indeed I would add it to .travis.yml but not worry about adding it in the requirements file (though you might add a note somewhere in the docs that scipy is required for the tests)

I don't have any strong preference for the module name, though hyphens always make it harder to have the package name patch the module name (i.e. one can't import astro-scrappy, only astro_scrappy).

cmccully commented:
@astrofrog Thanks. You make a good point. Can you please rename this repo without the dash then? I will update the name in the docs.

astrofrog commented:
@cmccully - done!

cmccully commented:
@astrofrog Thanks! Can you update travis so it does the build test? Also, should I uncomment the coveralls line in the travis file?

astrofrog commented:
@cmccully - yes, I have enabled coveralls, so you can enable it in the travis file. Ignore the latest failure, I think it was because of the repo name change. I think when you push the change to travis.yml it should work.

cmccully commented:
Any other comments before I pull this in?

larrybradley commented:
:+1: Thanks!

astrofrog commented:
:+1:

cmccully commented:
@astrofrog @larrybradley @cdeil @eteq @crawfordsm @mwcraig Thanks for all of your help with this. I am pulling in this request. I am going to wait a little longer before I release it on PyPi. Please feel free to continue testing this. Thanks again!

mwcraig commented:
Excellent! 

Once you release it on PyPI would you mind opening a pull request on https://github.com/astropy/conda-builder-affiliated to get it build as a conda package?
